Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -42,9 +55,238 @@ theorem radon_partition (h : ¬ AffineIndependent 𝕜 f) :
+  /- Use `Convex.radon_partition` to conclude there is a subset `I` of `ι` and a point `p : E`
+  which lies in the convex hull of either `a '' I` or `a '' Iᶜ`. We claim that `p ∈ ⋂₀ F`. -/
+  obtain ⟨I, p, h4_I, h4_Ic⟩ := Convex.radon_partition h_ind
+  use p
+  apply mem_iInter_of_mem
+  intro i
+  /- It suffices to show that for any set `F i` in a subcollection `I` of `ι`, that the convex
+  hull of `a '' Iᶜ` is contained in `F i`. -/
+  suffices ∀ I : Set ι, i ∈ I → (convexHull 𝕜) (a '' Iᶜ) ⊆ F i by
+    by_cases h : i ∈ I
+    · exact this I h h4_Ic
+    · apply this Iᶜ h; rwa [compl_compl]
+  /- Given any subcollection `I` of `ι` containing `i`, because `F i` is convex, we need only
+  show that `a j ∈ F i` for each `j ∈ Iᶜ`. -/
+  intro I hi
+  rw [Convex.convexHull_subset_iff (h_convex i)]
+  rintro - ⟨j, hj, rfl⟩
+  /- Since `j ∈ Iᶜ` and `i ∈ I`, we conclude that `i ≠ j`, and hence by the definition of `a`:
+  `a j ∈ ⋂ F '' (Set.univ \ {j}) ⊆ F i`. -/
+  apply mem_of_subset_of_mem (s₁ := ⋂₀ (F '' (Finset.univ.erase j)))
+  · apply sInter_subset_of_mem
+    use i
+    simp only [coe_erase, coe_univ, mem_diff, Set.mem_univ, mem_singleton_iff, true_and,
+      and_true]
+    exact fun h' ↦ hj (h' ▸ hi)
+  · apply Set.Nonempty.some_mem
+
+/-- **Helly's theorem** for finite families of convex sets in its classical form.
+
+If `F` is a family of `n` convex sets in a vector space of finite dimension `d`, with `n ≥ d + 1`,
+and any `d + 1` sets of `F` intersect nontrivially, then all sets of `F` intersect nontrivially. -/
+theorem helly_theorem {F : ι → Set E} [Fintype ι]
+    (h_card : finrank 𝕜 E + 1 ≤ Fintype.card ι)
+    (h_convex : ∀ i, Convex 𝕜 (F i))
+    (h_inter : ∀ I : Finset ι, I.card = finrank 𝕜 E + 1 → (⋂ i ∈ I, F i).Nonempty) :
+    (⋂ i, F i).Nonempty := by
+  apply helly_theorem' h_convex
+  intro I hI
+  obtain ⟨J, hJ_ss, hJ_card⟩ := Fintype.exists_superset_card_eq hI h_card
+  apply Set.Nonempty.mono <| biInter_mono hJ_ss (by intro _ _; rfl)
+  exact h_inter J hJ_card
+
+/-- **Helly's theorem** for finite sets of convex sets.
+
+If `F` is a finite set of convex sets in a vector space of finite dimension `d`, and any `k ≤ d + 1`
+sets from `F` intersect nontrivially, then all sets from `F` intersect nontrivially. -/
+theorem helly_theorem_set' {F : Finset (Set E)}
+    (h_convex : ∀ X ∈ F, Convex 𝕜 X)
+    (h_inter : ∀ G : Finset (Set E), G ⊆ F → G.card ≤ finrank 𝕜 E + 1 → (⋂₀ G : Set E).Nonempty) :
+    (⋂₀ (F : Set (Set E))).Nonempty := by
+  classical -- for DecidableEq, required for the family version
+  rw [show ⋂₀ F = ⋂ X : F, (X : Set E) by ext; simp]
+  apply Convex.helly_theorem' (F := fun x : F ↦ x.val) (𝕜 := 𝕜)
+  · exact fun X ↦ h_convex X (by simp)
+  · intro G _
+    let G' : Finset (Set E) := Finset.image Subtype.val G
+    rw [show ⋂ i ∈ G, ↑i = ⋂₀ (G' : Set (Set E)) by simp [G']]
+    apply h_inter G'
+    · rw [Finset.image_subset_iff]
+      simp
+    · apply le_trans Finset.card_image_le
+      assumption
```

Comment:
When one goal is clearly the main one/much longer than the rest, we tend to unindent it
```suggestion
  apply Convex.helly_theorem' (F := fun x : F ↦ x.val) (𝕜 := 𝕜)
  · exact fun X ↦ h_convex X (by simp)
  intro G _
  let G' : Finset (Set E) := Finset.image Subtype.val G
  rw [show ⋂ i ∈ G, ↑i = ⋂₀ (G' : Set (Set E)) by simp [G']]
  apply h_inter G'
  · rw [Finset.image_subset_iff]
    simp
  · apply le_trans Finset.card_image_le
    assumption
```